src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1013 @@
+#include "src/core/lib/address_utils/sockaddr_utils.h"
+#include "src/core/lib/event_engine/grpc_polled_fd.h"
+#include "src/core/lib/event_engine/nameser.h"  // IWYU pragma: keep
+#include "src/core/lib/event_engine/tcp_socket_utils.h"
+#include "src/core/lib/event_engine/time_util.h"
+#include "src/core/lib/gpr/string.h"
+#include "src/core/lib/gprpp/debug_location.h"
+#include "src/core/lib/gprpp/examine_stack.h"
+#include "src/core/lib/gprpp/host_port.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/gprpp/status_helper.h"
+#include "src/core/lib/iomgr/error.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+#include "src/core/lib/iomgr/resolved_address.h"
+#include "src/core/lib/iomgr/sockaddr.h"
+#ifdef GRPC_POSIX_SOCKET_ARES_EV_DRIVER
+#include "src/core/lib/event_engine/posix_engine/tcp_socket_utils.h"
+#elif defined(GRPC_WINDOWS_SOCKET_ARES_EV_DRIVER)
+#endif
+
+namespace grpc_event_engine {
+namespace experimental {
+
+grpc_core::TraceFlag grpc_trace_ares_wrapper_address_sorting(
+    false, "ares_wrapper_address_sorting");
+
+grpc_core::TraceFlag grpc_trace_ares_wrapper_stacktrace(
+    false, "ares_wrapper_stacktrace");
+
+#define GRPC_ARES_WRAPPER_STACK_TRACE()                                \
+  do {                                                                 \
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_wrapper_stacktrace)) { \
+      absl::optional<std::string> stacktrace =                         \
+          grpc_core::GetCurrentStackTrace();                           \
+      if (stacktrace.has_value()) {                                    \
+        gpr_log(GPR_DEBUG, "%s", stacktrace->c_str());                 \
+      } else {                                                         \
+        gpr_log(GPR_DEBUG, "stacktrace unavailable");                  \
+      }                                                                \
+    }                                                                  \
+  } while (0)
+
+grpc_core::TraceFlag grpc_trace_ares_wrapper(false,
+                                             "event_engine_ares_wrapper");
+
+namespace {
+
+EventEngine::Duration calculate_next_ares_backup_poll_alarm_duration() {
+  // An alternative here could be to use ares_timeout to try to be more
+  // accurate, but that would require using "struct timeval"'s, which just
+  // makes things a bit more complicated. So just poll every second, as
+  // suggested by the c-ares code comments.
+  return std::chrono::seconds(1);
+}
+
+bool IsIpv6LoopbackAvailable() {
+#ifdef GRPC_POSIX_SOCKET_ARES_EV_DRIVER
+  return PosixSocketWrapper::IsIpv6LoopbackAvailable();
+#elif defined(GRPC_WINDOWS_SOCKET_ARES_EV_DRIVER)
+  // TODO(yijiem): (debt) move pieces for Windows
+  return false;
+#else
+#error "Unsupported platform"
+#endif
+}
+
+struct HostbynameArg {
+  GrpcAresHostnameRequest* request;
+  const char* qtype;
+};
+
+constexpr char g_service_config_attribute_prefix[] = "grpc_config=";
```

Comment:
That's a good point. The config-parsing logic should move to the client channel resolver, which is the piece that is doing all the rest of the translation from DNS queries into service configs. Otherwise, if we keep this code as is, someone using `EventEngine::DNSResolver::LookupTXT` outside of gRPC is not going to get any results if they're looking for something in a TXT record other than a gRPC service config.

This wrapper code could simply make a string out of the response text and return it.

Reply:
Done. This is handled in https://github.com/grpc/grpc/pull/33030.